proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,429 @@
+   type) and ``Set.fromList``, the function. The fact that the module and type
+   have the same name is inconsequential here, but it is permitted.
+
+6. Suppose our library exports class ``C`` with method ``meth``. Our users
+   will frequently import ``C(..)`` and get ``meth`` in scope. Now I wish to
+   refactor ``C`` without changing my users' import behavior. There is no way
+   to do this currently.
+
+   With this proposal, we can do this::
+
+     -- top of file:
+     module MyLibrary ( module C(meth) ) where
+
+       class C a where ...
+
+       import module C where
+         meth :: ...
+
+   This example shows that modules may be *extended*. The ``class C``
+   declaration implicitly creates module ``C``, which is then extended below.
+
+   Import statements *do* have to change here: instead of ``C(..)``, they must
+   become ``import module C``. However, this new import statement is a drop-in
+   replacement for ``C(..)`` and may become preferable (as it is customizable
+   in the way demonstrated here).
+
+Proposed Change Specification
+-----------------------------
+
+1. Introduce a new extension ``-XLocalModules``.
+
+2. Introduce a new declaration form (allowed only at the top level of a
+   module) to declare new modules called *local modules*. Here is the BNF::
+
+     decl ::= ... | [ 'import' ] 'module' modname [ export_spec ] 'where' decls
+     modname ::= conid | '_'
+
+   Using ``_`` as the module name indicates an *anonymous local module*.
+
+   This declaration form is allowed only with ``-XLocalModules``.
+     
+3. All
+   declarations in scope at the declaration point of a local module (that is,
+   outside the module itself) are in scope for the entire local module.
+
+4. Definitions in a local module may be mutually recursive with definitions
+   in other local modules or outside of any local module. That is, local
+   modules influence scoping, but not type-checking or dependency.
+   
+5. A local module declaration brings into scope names listed in its export
+   list. These names are always brought into scope qualified by the local
+   module name, unless that module name is ``_``. If the declaration includes
+   the ``import`` keyword, the names are also brought into scope unqualified.
+
+6. It is an error to specify a non-empty export list, name a module ``_``, and
+   not include the ``import`` keyword. (Modules named ``_`` without an
+   non-empty export list but without the ``import`` keyword are still useful
+   as a way of declaring instances that use local definitions.)
+
+7. Every ``class``, ``data``, ``newtype``, ``data instance``, and ``newtype
+   instance`` declaration implicitly creates a new local module. The name of
+   the local module matches the name of the declared type. All entities (e.g.,
+   method names, constructors, record selectors) brought into scope within the
+   declaration, including the type itself, are put into this local module.
```

Comment:
This seems to be a nice convenience feature on top of the existing proposal, but is not crucial to it, right, as you can emulate that behavior with a few extra lines of wrapping the type in a module.

(Not that I dislike it, just asking for better understanding of the design space.)